### DISCUSSION TITLE
Dangerous behavior when jit-ting non-pure functions

### DISCUSSION THREAD

**Discussion by fischbacher** [jax-ml/jax#13415] [Q&A]
In a code review, the problem just came up that the author wrote code which accidentally mutated an input, but observed behavior looked OK since jax.jit changed semantics in such a way that this behavior did not surface.

The situation is equivalent to this example:

```python
import jax

def f(somedict):
  otherdict = {**somedict}
  otherdict['x'].setdefault('y', {})['z'] = 1
  return otherdict
  
jf = jax.jit(f)
```
```python
>>> d = dict(x=dict(y=dict(a=1)))
>>> d
{'x': {'y': {'a': 1}}}
>>> f(d)
{'x': {'y': {'a': 1, 'z': 1}}}
>>> d
{'x': {'y': {'a': 1, 'z': 1}}}
>>> 
>>> d2 = dict(x=dict(y=dict(a=1)))
>>> jf(d2)
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
{'x': {'y': {'a': DeviceArray(1, dtype=int32, weak_type=True), 'z': DeviceArray(1, dtype=int32, weak_type=True)}}}
>>> d2
{'x': {'y': {'a': 1}}}
```

So, when calling this `f()` on a dict `d` that has a dict entry for key `'y'`, we are returning a copy of `d`, but when amending this copy, we actually mutate the dict-object that also is the value of `f['y']` - so, we mutate nested content of `d`.

In general, when people do this, this is a nasty bug. It's insiduous since "calling X mutates an argument in an unexpected way" easily causes behavior that's of the "a change in one place makes something blow up in an entirely unrelated place".

Now, the jitted `f` has different behavior. Arguably, the JAX documentation (https://jax.readthedocs.io/en/latest/_modules/jax/_src/api.html#jit) states that the function we are jitting must be pure, so we are actually violating the JAX API contract here and hence "cannot expect anything". The concern here is that authors write code, potentially might even see a possible concern about accidentally mutating an input argument, try it out with the jitted function, and conclude that everything is OK. In that situation, code would be relying on undocumented JAX behavior, and simply removing the jitting would introduce a serious bug.

Would it be feasible to bolster detection of non-pure behavior of a jax.jit-transformed function to replace "you are relying on undocumented behavior" with "this raises an exception".

**Comment by jakevdp** [comment]
Thanks for the report – it seems like this same issue can be demonstrated with a simpler function:
```python
from jax import jit
def f(d):
  d['z'] = 1

d = {}
f(d)
print(d)
# {'z': 1}

d = {}
jit(f)(d)
print(d)
# {}
```
As you mention, JAX transforms assume that functions are pure, and functions like this violate that contract (see [Sharp Bits: Pure Functions](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions) for discussion).

JAX does have some guardrails to detect similar effectful code using the `jax_check_tracer_leaks` setting. It does not detect the original issue you bring up because the code doesn't mutate inputs under JIT, but it will detect similar issues. For example:
```python
import jax
jax.config.update('jax_check_tracer_leaks', 1)
global_dict = {}

def f(x):
  global_dict['x'] = x

jax.jit(f)(1)
# Exception: Leaked sublevel 1. Leaked tracer(s):

# Traced<ShapedArray(int32[], weak_type=True)>with<DynamicJaxprTrace(level=0/1)>
# This DynamicJaxprTracer was created on line <ipython-input-14-807856f72959>:8 (<module>)
# <DynamicJaxprTracer 139832618947408> is referred to by <tuple 139832974830224>[0]
```
We could think about how to detect your situation under the same flag, but at first glance I'm not entirely sure how we'd do that. It's probably possible with some thought... Could you open a feature request linking to this discussion? Issues can be assigned, so they're less likely to get lost in the mix.
